### ISSUE TITLE
Intersection of iterables should contain elements that are intersection of iterable type params

### ISSUE THREAD

**Issue by weirdan** [vimeo/psalm#710]
https://getpsalm.org/r/d7df39b6dd
```php
<?php
  class A {} function takesA(A $p): void {}
  class B {} function takesB(B $p): void {}
  /** @psalm-param iterable<A>&iterable<B> $i */
  function takesIntersectionOfIterables(iterable $i): void {
    foreach ($i as $c) {
      takesA($c);
      takesB($c); // fails with `A provided where B expected`
    }
  }
  /** @psalm-param iterable<A&B> $i */
  function takesIterableOfIntersections(iterable $i): void {
    foreach ($i as $c) {
      takesA($c);
      takesB($c); // does not fail
    }
  }
```
Currently psalm complains that `takesB($c)` is invalid (`$c` is treated as being *only* `A`). But `iterable<A>&iterable<B>` is equivalent to `iterable<A&B>` (and that does not complain)

**Comment by weirdan** [comment]
Similarly, psalm currently does not accept the following:
```php
  /** @psalm-var iterable<A>&iterable<B> $data */
  $data = [];
  takesIterableOfIntersections($data); // defined above
```

**Comment by muglug** [comment]
I'm not *entirely* sure this is a bug, but I agree that Psalm should probably rewrite `iterable<A>&iterable<B>` to `iterable<A&B>` as a convenience.
